xwiki-platform-core/xwiki-platform-model/xwiki-platform-model-api/src/main/java/org/xwiki/model/internal/reference/AbstractReferenceEntityReferenceResolver.java

```diff
@@ -90,18 +90,19 @@ private EntityReference normalizeReference(EntityReference referenceToResolve, O
         EntityReference reference = normalizedReference;
         while (reference != null) {
             List<EntityType> types = reference.getType().getAllowedParents();
-            if (reference.getParent() != null && !types.isEmpty() && !types.contains(reference.getParent().getType())) {
+            if (reference.getParent() != null
+                && isParentTypeAndAllowedTypeNotMatching(types, reference.getParentType())) {
                 // The parent reference isn't the allowed parent: insert an allowed reference
                 EntityReference newReference =
                     resolveDefaultReference(types.get(0), parameters).appendParent(reference.getParent());
                 normalizedReference = normalizedReference.replaceParent(reference.getParent(), newReference);
                 reference = newReference;
-            } else if (reference.getParent() == null && !types.isEmpty()) {
+            } else if (reference.getParent() == null && reference.getParentType() != null) {
                 // The top reference isn't the allowed top level reference, add a parent reference
-                EntityReference newReference = resolveDefaultReference(types.get(0), parameters);
+                EntityReference newReference = resolveDefaultReference(reference.getParentType(), parameters);
                 normalizedReference = normalizedReference.appendParent(newReference);
                 reference = newReference;
-            } else if (reference.getParent() != null && types.isEmpty()) {
+            } else if (reference.getParent() != null && (types.isEmpty() || reference.getParentType() == null)) {
```

Comment:
`reference.getParentType()` cannot be null if `reference.getParent()` is not null

Reply:
right

Reply:
> reference.getParentType() cannot be null if reference.getParent() is not null

actually that's not true: it can be null if `reference.getParent()` is a wiki.

Reply:
> actually that's not true: it can be null if reference.getParent() is a wiki.

If `reference.getParent()` is a wiki then `reference.getParentType()` returns WIKI.